### PR TITLE
Improve ExchangeMessageDispatch

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
There are a few TODO left over ExchangeManager and ExchangeContext

#### Change overview
* Inline `ExchangeMessageDispatch::SendMessage` and `ExchangeMessageDispatch::OnMessageReceived`, as they are only used once, and they are non-virtual
* Pull `dispatch.IsEncryptionRequired()` check at exchange creation, do not run the check on every message.

Note:
* The group session check is removed because it is inherited by `mSession->RequireMRP()` check.
* The encryption check in `ExchangeContext::MatchExchange` is removed because it is inherited by session match.

#### Testing
Passed unit-tests.

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: src/messaging/ExchangeMgr.cpp [hunk 1]
```diff
@@ -288,6 +288,12 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
             }
         }
 
+        if (GetDispatchForDelegate(delegate).IsEncryptionRequired() != session->IsEncrypted())
+        {
+            ChipLogError(ExchangeManager, "OnMessageReceived failed, err = %s", ErrorStr(CHIP_ERROR_NO_UNSOLICITED_MESSAGE_HANDLER));
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.cpp:293]: That's not the error type we used to have here.  Why is this the right type?

Also, this check will _always_ fail if `delegate` is null, right?  As long as we are moving this around, can we skip this check for the `!delegate` case where we're just trying to send a standalone ack?  Or will something go terribly wrong if the session is not in fact encrypted and we try to send an ack via ApplicationExchangeDispatch?

**Comment by kghost** [comment]
I incline removing `ExchangeMessageDispatch` completely. After this PR there will be only 2 methods in this class:

```c++
    virtual bool MessagePermitted(uint16_t protocol, uint8_t type) = 0;
    virtual bool IsEncryptionRequired() const { return true; }
```

Only `CASESession`/`PASESession` utilize these function, so we can definitely move those method into `CASESession`/`PASESession` class, and remove `ExchangeMessageDispatch` class.

This approach is more memory efficient, less object to maintain, more robust and readable. @bzbarsky-apple any opinions ?

**Comment by kghost** [comment]
Will split this PR into several small PRs

**Comment by bzbarsky-apple** [comment]
I don't think it's just CASESession/PASESession that use IsEncryptionRequired...

Past that, I don't have any particularly strong views about the way this ends up set up as long as we are doing equivalent checks with equivalent safety guarantees.

**Comment by bzbarsky-apple** [comment]
But also: this seems like it should be a much lower priority than things like fixing regressions like #16748 or other crash bugs like #15305.  Unless this is blocking 1.0 in some way, we probably should not be rejiggering things that work at the moment.

@andy31415

**Pull request closed without merging**